[PR-1256] Existential vs universal quantification (#4)
serokell/blog-posts: existential-vs-universal-quantification/article.md

# Existential vs universal quantification

> ## Introduction

Reviewer: Would add some information for why the person reading it would be interested in the article in the front.

From type theory we know about existence of two quantifiers - `$\forall$` and `$\exists$`.
The former means: `"For every term x the statement is true"`, the later: `"There is at least one term x for which this statement is true"`.

Since Haskell is based on type theory there is a place for these quantifiers in language syntax, however, in Haskell we are talking not about some terms and statements, but about types. 

> By default, all type variables are bounded by `forall` quantifier, which means that every type, that passes constraints, can be used inside this function/type declaration.

Reviewer: Would consider giving an example

You can't see this quantification without using a specific extension - "Explicit Forall".

> ## Universal quantification and ExplicitForall

Reviewer: Perhaps there can be a bit clearer structure. 

## Universal quantification 

(Explain what it is)

### Where can it be used? 

### Example: x 

## Existential quantification 

(Explain what it is)

### Where can it be used? 

### Example: y 

## Some cool places where people can learn more

The mentioned extension is enabled by using this pragma:
```haskell
{-# LANGUAGE ExplicitForAll #-}
```
It allows us to use `forall.` quantifier in functions and types declarations manually.
By default, `forall.` is implicit and sometimes it can lead to unpleasant situations.
Consider this code:

```Haskell
prepend2 :: a -> [a] -> [a]
prepend2 x xs = pairFun x ++ xs
 where pairFun y = [y, y]
```
This code compiles without any errors, until you try to use variables from the outer function inside the `pairFun` function.
```Haskell
-- code compiles with error
prepend2 :: a -> [a] -> [a]
prepend2 x xs = pair ++ xs
 where pair :: [a]
       pair = [x, x]
```

If you try to compile this one you will get an error because compiler can't match type variable `a` from outer
declaration, with the one in inner declaration. 
This is happening, because the real signature of this function is this:

```Haskell
prepend2 :: forall a. a -> [a] -> [a]
prepend2 x xs = pair ++ xs
 where pair :: forall a. [a]
       pair = [x, x]
```

In order to correct the code you need to use `-XScopedTypeVariables` extension, which will only work if you
explicitly add forall keyword to `prepend2` function.

> ##### Usage

Reviewer: ##, then ###, then ####

> Let's try to add `forall.` manually.

Reviewer: Do you need to repeat that they need to add {-# LANGUAGE ExplicitForAll #-}? People might have missed.

However, using explicit `forall.` this can be simplified:

```Haskell
veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
veryLongFunction = ...

func = veryLongFunction @Integer ...
```

Since `j` is the first type variable in the declaration of `veryLongFunction` you can specify only it.
This is useful not only in such big examples, but also when your function requires any of its arguments to be specified on usage.

##### Additional type variables
Second, you now can define functions with more type variables than you use on the right side of declaration.
This is used mostly when you want to add complex constraint, which require several types, without adding them into the definition itself:

``` Haskell
someMonadicFunction :: forall a m env. MonadReader env m => a -> m a
```
 
Here we don't have `env` used in the type definition, but despite that, we still able to add `MonadReader` constraint.

## Existential quantification
Beside universally quantified variables, Haskell also supports existential types. 
In order to use them ... you still need to use `forall.` keyword ¯\\\_(ツ)_/¯.
That might be a little confusing.

> The following construction is called existential type in Haskell - when you use forall quantifier in the right-hand side of function or datatype declaration. 

Reviewer: I would explain here why it results in an existential type, and how it relates to existential quantification, think that would be interesting.

But what is it used for? Let's find out!
We will start with functions:

```haskell
func :: forall a c. a -> (forall b. (Show b) => b) -> c
```
Few things to mention.
Here `a` and `c` are universally quantified and `b` is existentially quantified variables. 
The same variable can't be quantified more than once.
Also, you can see that constraints for `b` are added in the middle of the function type declaration.
This example is pretty useless.
The only thing that has changed here is that now you can't manually choose type for `b` - the compiler must deduce the correct variable you want to use by itself.

When we come to datatype declaration, it becomes more useful. 

```Haskell
data MyData a = forall b. MkMyData b a
```
The first thing you might have noticed is that we don't have type variable for `b` on the left side of expression.
However, we still can use it on the right side.
This is because `b` is existentially quantified.

```Haskell
MkMyData @_ @Integer -- This won't work, you don't have second variable in declaration
```
Compiler will deduce the type of the variable itself, and you can't specify it. 
It allows you to introduce variables, without need to specify them.

> This might be useful for example in this case:

Reviewer: substituting "for example, in this case:" with "when building a multi-typed list" would make it easier to understand I think

# Existential vs universal quantification

## Introduction
This article gives brief information about one of the most common extension in Haskell projects - `ExplicitForAll`

> and syntax constructions that this extension brings to language.

Reviewer 2: ```suggestion
This article gives brief information about one of the most common extension in Haskell projects, `ExplicitForAll`,
and the syntax constructions that this extension brings to language.
```

Stay tuned if you want to know how to create lists of differently typed variables, how to add complex constraints and more!

> From type theory we know about the existence of two quantifiers - `$\forall$` and `$\exists$`.

Reviewer 2: I don't know anything about type theory, but I think the forall/exists quantifiers come from predicate logic: https://en.wikipedia.org/wiki/First-order_logic

Reviewer 2: But if you want to mention the origins of this (whether it be type theory or predicate logic or something else), it might be sensible to ask someone more knowledgeable on the topic to review the introduction.

> The former means: `"For every term x the statement is true"`, the later: `"There is at least one term x for which this statement is true"`.

Reviewer 2: > The former means: `"For every term x the statement is true"`

There is no mention of `x` in the paragraph above, i.e.:

> From type theory we know about the existence of two quantifiers - `$\forall$` and `$\exists$`.

For a reader who's not yet familiar with the topic (aka the target audience), I suspect this will be a bit confusing.

> Since Haskell is based on type theory there is a place for these quantifiers in language syntax, however, in Haskell we are talking not about some terms and statements, but about types.

Reviewer 2: Since this paragraph (and the paragraphs below it) intends to explain what universal quantification is, should it go under the `## Universal quantification and ExplicitForall` header?

> Since Haskell is based on type theory there is a place for these quantifiers in language syntax, however, in Haskell we are talking not about some terms and statements, but about types.

Reviewer 2: ```suggestion
Since Haskell is based on type theory there is a place for these quantifiers in language syntax, however, in Haskell we are not talking about some terms and statements, but about types.
```

> By default, all type variables are bound by a `forall` quantifier, which means that every type, that passes constraints, can be used inside this function/type declaration.

Reviewer 3: > every type, that passes constraints, can be used inside this function/type declaration

This might be slightly misleading, as it could give the impression that the function is free to choose what the type `a` can be, inside it. But AFAIU, you cannot take this argument, and say, use it as a `String`, or an `Int` or any other concrete type.

Reviewer 2: Agreed, it's important to understand the caller/callee dichotomy.

With universal quantification, the **caller** can choose how to instantiate the type variable, but the **callee** can't.

With existential quantification, the **callee** can choose how to instantiate the type variable, but the **caller** can't.

We should try to make this distinction abundantly clear.

> By default, all type variables are bound by a `forall` quantifier, which means that every type, that passes constraints, can be used inside this function/type declaration.

Reviewer 2: I think the `, that passes constraints,` bit can be removed. Constraints are not related to universal/existential quantifiers.

----

In fact, depending on how you look at it, constraints can be viewed as quantifiers in their own right, see [our other blog post](https://serokell.io/blog/how-dependent-haskell-can-improve-industry-projects#quantifiers-and-their-attributes):

> Another example of a quantifier is ctx => ..., which introduces a class constraint.

```haskell
id :: forall a. a -> a
id x = x
```

> You can't see this quantification without using a specific extension - "Explicit Forall".

Reviewer 3: It is a bit unclear what you mean by "see" here. Also I think the space in "Explicit Forall" is not required.

Reviewer 2: Agreed. The extension does not let you _see_ the quantifier, it lets you _write_ it explicitly. Perhaps it should be rephrased to something like:

> In order to make the quantifier explicit, rather than implicit, you need to use the `ExplicitForAll` extension.

> ## Universal quantification and ExplicitForall

Reviewer 2: (Perhaps unintuitively) the `a` in the extension name must be upper case: `ExplicitForAll`, not `ExplicitForall`.

Also added the missing backticks.

```suggestion
## Universal quantification and `ExplicitForAll`
```

> ## Universal quantification and ExplicitForall

Reviewer 2: The `Universal quantification and ExplicitForAll` section seems a bit unstructured and doesn't seem to have a central narrative.
Let's write down a quick summary of what we have:

* Introduction: a single sentence explaining what universal quantification is + an example (`id`)
  * Universal quantification and ExplicitForAll: an example showing a possible use case for `ExplicitForAll` (`prepend`)
    * Usage: an example explaining the syntax (`func`)
    * Examples: 2 more examples showing more use cases for `ExplicitForAll` again

Here are some suggestions:

1. As stated in [my other comment](https://github.com/serokell/blog-posts/pull/4/files#r782087623), the sentence explaining what universal quantification is should be in the the `Universal quantification and ExplicitForAll` section, not in the introduction.
1. The `Universal quantification and ExplicitForAll` section should focus more on _what_ universal quantification is, and when and why it's useful, and also mention _who_'s responsible for instantiating universally quantified type variables (see [my comment here](https://github.com/serokell/blog-posts/pull/4/files#r782085135)). I think it's important to establish that _before_ mentioning that (1) there's an implicit `forall` and (2) it can be made explicit. The subsections can then go into detail about _why_ and _when_ a user would want to make it explicit.
1. We're explaining the syntax near the end of the section, on our 3rd example, `func`. By this point, the reader is already familiar with the syntax.
   So maybe we should explain the syntax before diving into complex examples. 
1. Looks like the goal of the `prepend` example is to show _when_ making the quantifier explicit is needed. 
   This is the same goal as `veryLongFunction` and `monadicFunction`, so looks like `prepend` belongs in the `Examples` section.

Reviewer: Seems like the rewrite addresses most of this comment so I'll resolve it

By default, `forall.` is implicit and sometimes it can lead to unpleasant situations.
Consider this code:

```Haskell
prepend2 :: a -> [a] -> [a]
prepend2 x xs = pairFun x ++ xs

> where pairFun y = [y, y]

Reviewer 2: These code snippets are not syntactically valid, the `where` clause needs indentation.

> where pairFun y = [y, y]

Reviewer 2: The helper function is named `pairFun` here, and just `pair` in the the snippet below.

We should settle on a single name (I suggest `pair`)

```
This code compiles without any errors, until you try to use variables from the outer function inside the `pairFun` function.
```Haskell
-- code compiles with error
prepend2 :: a -> [a] -> [a]
prepend2 x xs = pair ++ xs
where pair :: [a]
      pair = [x, x]
```

If you try to compile this one you will get an error because compiler can't match type variable `a` from outer
declaration, with the one in inner declaration.

> This is happening, because the real signature of this function is this:

Reviewer 2: I think "top-level definition" and "local definition" are more standard names than "outer/inner" (but I'd ask for more opinions on this)

```suggestion
If you try to compile this one you will get an error because the compiler can't match type variable `a` from the top-level
declaration with the one in the local declaration.
This happens because the real signature of this function is this:
```

```Haskell
prepend2 :: forall a. a -> [a] -> [a]
prepend2 x xs = pair ++ xs
where pair :: forall a. [a]
      pair = [x, x]
```

> In order to correct the code you need to use `-XScopedTypeVariables` extension, which will only work if you

Reviewer 2: _Why_ does adding `ScopedTypeVariables` solve the issue?
If we're gonna mention `ScopedTypeVariables`, we should also mention what it does and why it solves this problem.

---

I think this `prepend` example is good for demonstrating the need for `ScopedTypeVariables`, but that's beside the point of this blog post.
I don't think this example is good for demonstrating the need for `ExplicitForAll`. I suggest we come up with a different, more to-the-point example.

Reviewer 2: Done: https://github.com/serokell/blog-posts/pull/4/commits/aabbf1fb65912d7ee8dcfc2f2698b6f8a4f52719

> In order to correct the code you need to use `-XScopedTypeVariables` extension, which will only work if you

Reviewer 2: In this article, we use the `-X` flag prefix for the `ScopedTypeVariables` extension, but not for other extensions.

I think we should be consistent and not use `-X` anywhere.

> explicitly add forall keyword to `prepend2` function.

Reviewer 2: Why is the function called `prepend2` and not `prepend`?

### Usage
Let's try to add `forall.` manually.
The syntax is very simple - at the beginning of function or instance declaration, before any constraints or arguments were introduced use

> quantifier with all type variables you want to introduce later (Note: you get a compilation error if you use undeclared

Reviewer 2: * introduced -> used. In `a -> b` we're _using_ type variables. In `forall a b.` we're _introducing_ type variables.

```suggestion
The syntax is very simple - at the beginning of function or instance declaration, before any constraints or arguments are used, use
the quantifier to introduce all type variables you want to use later (Note: you get a compilation error if you use undeclared
```

type variables after forall):

```Haskell
func :: forall a b c m. a -> b -> m c -- OK, it compiles
func :: forall a b c.   a -> b -> m c -- Error, type variable `m` is not introduced
```
Here we introduce four variables. We might want to introduce some constraints.

> For example, it is very likely from the function definition that `m` is an instance of `Monad` typeclass.

Reviewer 3: > it is very likely from the function definition that `m` is an instance of `Monad` typeclass.

But why a `Monad` ?  Doesn't the signature only tell us that `m` is a type that takes another type as a paramter/argument?

Author: Well, when I see an type variable named m as the wrapper over the result, I subconsciously think of it as of some monad. But you are right, I think I need to correct it

In order to add this constraint we need to place it after all variables are declared:

```Haskell
func :: forall a b c m. Monad m => a -> b -> m c
``` 
You might want to ask, what is so special about this syntax?
Nothing changed since we added the quantifier.
Well, actually there are few things to notice here.

### Examples
#### Order of type variables
First of all, now you can change the order of type variables.
Let's imagine you want to define a function with 10 type variables (Yeah, there are real life examples of such functions), and you know that when you use it, you want to specify the last argument explicitly.

> Without forall you will write something like this:

Reviewer 3: I think we should at least mention about the `TypeApplication` extension as part of this section.

```Haskell
veryLongFunction :: a -> b -> c -> d -> e -> f -> g -> h -> i -> j
veryLongFunction = ...

func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
``` 

Quite long right?
This is because ghc will maintain the same order, as these variables first appear in function definition.
However, using explicit `forall.` this can be simplified:

```Haskell
veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
veryLongFunction = ...

func = veryLongFunction @Integer ...
```

Since `j` is the first type variable in the declaration of `veryLongFunction` you can specify only it.
This is useful not only in such big examples, but also when your function requires any of its arguments to be specified on usage.

#### Additional type variables

> Second, you now can define functions with more type variables than you use on the right side of declaration.

Reviewer 2: We don't need to make the quantification explicit in order to do this....

This is used mostly when you want to add complex constraint, which require several types, without adding them into the definition itself:

``` Haskell
someMonadicFunction :: forall a m env. MonadReader env m => a -> m a
```

> Here we don't have `env` used in the type definition, but despite that, we are still able to add `MonadReader` constraint.

Reviewer 2: You're implying that the constraints and other quantifiers are not part of the type definition, which I would argue is not true.

## Existential quantification
Beside universally quantified variables, Haskell also supports existential types.
In order to use them ... you still need to use `forall.` keyword ¯\\\_(ツ)_/¯.
That might be a little confusing.
The following construction is called existential type in Haskell - when you use a forall quantifier in the right-hand side of function or data type declaration.
But what is it used for? Let's find out!

### Usage

We will start with functions:

```haskell

> func :: forall a c. a -> (forall b. (Show b) => b) -> c

Reviewer 2: This requires `RankNTypes` in order to compile

```
Few things to mention.
Here `a` and `c` are universally quantified and `b` is existentially quantified variables.
The same variable can't be quantified more than once.
Also, you can see that constraints for `b` are added in the middle of the function type declaration.

> This example is pretty useless.

Reviewer 2: Are you saying that this example in particular is useless or that existential type variables in function declarations (rather than in data declarations) are useless?

Either way, if this example is useless, we should come up with something more realistic that better demonstrates its usefulness.

Off the top of my head, here's a motivating example:

You start out with this:
```hs
func :: IO ()
func = do
  username <- getUsername
  print username
  
  userCount <- getUserCount
  print userCount

getUsername :: IO String
getUsername = undefined

getUserCount :: IO Int
getUserCount = undefined
```

But then you realize you want to be able to choose whether `func` should log to the console or to a logfile. So you want to parameterize `func` over _some_ log function, which may or may not be `print`.

So, you need `func` to take in a log function capable of logging _anything that is showable_ (could be an int, a string, a timestamp, whatever, as long as it's showable). The **caller** of `func` does not know what will be logged.

For that, you need to existentially qualify the type variable `a`.

```hs
func :: (forall a. Show a => a -> IO ()) -> IO ()
func log = do

  username <- getUsername
  log username
  
  userCount <- getUserCount
  log userCount
```

```hs
main :: IO ()
main = do
  -- log to the console
  func print

  -- log to a logfile
  func (appendFile "logfile.log" . show)
```

IOW, `func` (the **callee**) is the one who decides how to instantiate `a`, not `main` (the **caller**)

The only thing that has changed here is that now you can't manually choose type for `b` - the compiler must deduce the correct variable you want to use by itself.

When we come to data type declaration, it becomes more useful.

#### Hiding type variables
```Haskell

> data MyData a = forall b. MkMyData b a

Reviewer 2: This doesn't compile as-is. It needs an extension to compile, but you haven't mentioned which.

Please bear in mind that, while we tend to enable a bunch of extensions by default in our projects, the reader might not. We should assume the reader has no extensions enabled. And if an extension is needed to compile our code, we should tell them.

```
The first thing you might have noticed is that we don't have a type variable for `b` on the left side of expression.
However, we still can use it on the right side.
This is because `b` is existentially quantified.

```Haskell

> MkMyData @_ @Integer -- This won't work, you don't have second variable in declaration

Reviewer 2: > -- This won't work, you don't have second variable in declaration

Not sure what you mean by this, it seems to work fine:

```hs
res :: MyData Integer
res = MkMyData @_ @Integer 1 1
```

The `MkMyData` constructor has two universally quantified type variables:

```
λ> :set -fprint-explicit-foralls
λ> data MyData a = forall b. MkMyData b a
λ> :t +v MkMyData
MkMyData :: forall a b. b -> a -> MyData a

```
Compiler will deduce the type of the variable itself, and you can't specify it.

That is actually why this construction is called existential type - because you tell a compiler, that there exists type, that

> can be used in constructor, without specifying which one exactly. Also it is called so, because from type theory's point of view

Reviewer 2: > because you tell a compiler, that there exists type, that
> can be used in constructor, without specifying which one exactly

I'm not sure this is very elucidating.

If had written the code below, I would also be telling the compiler that "there exists a type that can be used in the constructor, without specifying which one exactly":

```hs
data MyData a b = MkMyData a b
```

this constructions are isomorphic:

```haskell
data MyTypeWithFunc = forall a. MkMyTypeWithFunc a (a -> String)

MkMyTypeWithFunc :: forall a . a -> (a -> String)-> MyTypeWithFunc

> MkMyTypeWithFunc :: (exists a . (a, a -> String)) -> MyTypeWithFunc

Reviewer 2: Here you're introducing a keyword (`exists`) without saying what it is, what it does, or that it doesn't exist in haskell.

And saying that those two things are isomorphic doesn't seem to convey any information to the reader. What are they supposed to learn from this snippet?

```

#### Heterogeneous lists

Now we understand that existential types allow you to introduce variables, without necessity to specify them.

> This might be useful when building a multi-typed list:

Reviewer 2: _Why_ would the user want to do this? The `Elem` example is not very motivating, there's literally nothing you can do with a value of type `Elem`. So having a list of `[Elem]` is also not useful.

Perhaps a more motivating example would be to create a list of things that can be displayed on the screen:

```hs
data Showable = forall a. Show a => MkShowable a

printMany :: [Showable] -> IO ()
printMany xs = 
  forM_ xs $ \(MkShowable x) -> print x
```

```
λ> printMany [MkShowable 1, MkShowable "hello"]
1
"hello"
```

```haskell
data Elem = forall a. Elem a

multitypedList :: [Elem]
multitypedList = [Elem "a", Elem 1, Elem (Just 5)]
```

> This is a heterogeneous list that contains variables of different types.

Reviewer 3: > contains variables

I think it might be better to say "value" instead of "variable".

> You couldn't do that without existential types, because even if you use type variables, you should still have every element in the list have the same type.

Reviewer 2: > because even if you use type variables

You're implying that in this code snippet we're not using type variables, but we are. It's just that it's existentially quantified rather than universally.

Note, that since type variable `a` is hidden with quantifier, you can't add constraints to it, beside those added to the constructor

```haskell
showFunc :: Show ??? => [Elem] -> String -- There is no type variable, that you can add constraint to
showFunc (x:xs) => (show x) ++ (showFunc xs)
```

> The same "shadowing" is used in well-known `SomeException` wrapper:

Reviewer 2: What do you mean by "shadowing" in this context?

Author: By shadowing I meant hiding types of fields. Maybe it has a more appropriate term?

```
This is a heterogeneous list that contains values of different types.
You couldn't do that with universally quantified list, because in that case you should have every element in the list have the same type.
There are not many things you can do with elements of this list.
Since their type is unknown, you can't address it, nor can you pass it to functions that expect values of concrete type.
However, adding constraints to `a` variable may improve the situation

```haskell
data Elem = forall a. (Show a) => Elem a

multitypedList :: [Elem]
multitypedList = [Elem "a", Elem 1, Elem (Just 5)]

printElem :: Elem -> IO ()
printElem (Elem x) = print x

main :: IO ()
main = fmap printElem multitypedList

λ> "a"
λ> 1
λ> Just 5
```

Note, that since type variable `a` is hidden with quantifier, you can't add constraints to it, beside those added to the constructor

```haskell
showFunc :: Show ??? => [Elem] -> String -- There is no type variable, that you can add constraint to
showFunc (x:xs) => (show x) ++ (showFunc xs)
```

The same way we hide exception type in well-known `SomeException` wrapper:
```haskell
data SomeException = forall a. Exception a => SomeException a
```

> It allows you to catch exceptions of any type (presumed it was wrapped inside `SomeException`) and then check if this is an exception of the type you want.

Reviewer: I don't really understand what presumed means here.

## Summary

We had taken a look at quantification in Haskell.
While this hadn't introduced many syntax constructions, it allows you to further expand rich type system of the language,
and to create constructions, which previously were impossible.

> This extension is often used in haskell projects (I actually hadn't yet found any big project without it), so it is very

Reviewer: This sentence should be moved somewhere to the beginning

# Existential and Universal Quantification in Haskell

This article gives a brief introduction into one of the most common Haskell extensions – `ExplicitForAll` – and syntax constructions that this extension brings to language.

> I haven't yet found any big project that wouldn't use this extension, so it is very

Reviewer 4: Tone: first person vs academic-neutral

Reviewer: Yeah, cool catch. 👌

I'd use first person: In this article, I give a brief introduction..

Author: done

useful to either become acquainted with it or to refresh it in your memory if you already know about it.

Stay tuned if you want to know how to create lists of differently typed variables, how to add complex constraints, and more!

## Introduction

From predicate logic, we know about the existence of two quantifiers – `$\forall$` and `$\exists$` – that we can apply to a term `x` and some statement.

> The former means: "For every term x, the statement is true", the later: `"There is at least one term x for which this statement is true"`.

Reviewer 2: ```suggestion
The former means: "For every term x, the statement is true", the later: "There is at least one term x for which this statement is true".
```

> The former means: "For every term x, the statement is true", the later: `"There is at least one term x for which this statement is true"`.

Reviewer 4: Unnecessary backticks

Author: done

> Since Haskell is based on predicate logic, there is a place for these quantifiers in language syntax. However, in Haskell, we are talking not about some terms and statements but about types.

Reviewer 2: > Since Haskell is based on predicate logic

Do you have a source for this assertion?

Author: Haskell based on Hindley–Milner type inference which is based on lambda calculus which is based on term - 'Term', which comes from firs order predicate logic (TBH, I am not sure if it comes from predicate logic, or something deeper)

Reviewer 4: FOPL is pretty atomic as far as PLT is concerned :)

> Since Haskell is based on predicate logic, there is a place for these quantifiers in language syntax. However, in Haskell, we are talking not about some terms and statements but about types.

Reviewer 4: You're kinda explaining forall by claiming that we are based on forall in the first place. I would reword to something like

"consider signature of id. Actually, we can add forall a. a -> a without breaking anything. This is an illustration of the fact that Haskell type inference is rooted in the predicate logic".

```haskell
id :: forall a. a -> a
id x = x
```

> This quantification is implicit by default (if you check the type signature of `id`, there is no quantifier there). To make it explicit, you need to use a specific extension – "ExplicitForAll".

Reviewer 4: As I said, in this section, invert the narration. We already know that id :: a -> a, tell us that it's also forall a. a -> a, then say that it illustrates the roots of HMTI.

Author: How would you mantion HMTI here, after showing true `id` type? I thought about something like
"This quantifiers come from Hindley–Milner type system ...", but I don't know what to add instead of ...

## Universal quantification and `ExplicitForAll`

> The key point of both universal and existential quantification is choosing who is responsible for deducing types of a function, datatype, or class.

Reviewer 2: > who is responsible for deducing types

By "deducing types", did you mean inferring types?

Universal & existential quantification are not about type inference, it's about type instantiation (i.e. choosing a type for a type variable).
Type inference is just one of two ways of instantiating types.

When you apply a function with type variables, you can either let GHC infer how to instantiate the type variables:

```hs
add :: Num a => a -> a -> a

-- GHC will infer that `a` should be instantiated to `Integer`
result = add 1 2

-- GHC will infer that, since the expression `add 1 2` is of type `Int`, then `a` must also be an `Int`
result = add 1 2 :: Int
```

Or you could disable type inference, and explicitly tell GHC to instantiate `a` to `Int`.

```hs
add @Int 1 2
```

So, as you can see, type inference is not related to the topic.

It doesn't matter _how_ variables get instantiated (via inference or via type applications), what matters is who's responsible for instantiating them.

By default, all type variables are considered universally quantified.

> It means that the **caller** of a function (so, the programmer, actually) is responsible in case the type variable is ambiguous.

Reviewer 2: > the **caller** of a function (so, the programmer, actually) 

> In contrast to the universal quantifier, the existential one means that the compiler (**callee**) is responsible for deducing what a type variable should mean 

Like I mentioned in my other comment: a type variable can be inferred by the compiler or explicitly instantiated by the programmer; it doesn't matter whether it's universally or existentially quantified.

------

The distinction I was trying to make between caller & callee (in [slack](https://serokell.slack.com/archives/CBC4ZN26R/p1642443064015700?thread_ts=1642440140.014200&cid=CBC4ZN26R), [in this comment](https://github.com/serokell/blog-posts/pull/4#discussion_r782411289) and [in this gist](https://gist.github.com/dcastro/321fb951d768907f61598c4ce30906ab)) was not a distinction between programmer & compiler.

Perhaps the (very **informal**) terms I used (caller & callee) were not the best to describe what I meant.
Maybe "use-site" and "definition-site" would be better?
Let me try another way. Let's reuse the examples from my [gist](https://gist.github.com/dcastro/321fb951d768907f61598c4ce30906ab).

```hs
-- `a` is universally quantified
head :: forall a. [a] -> Maybe a

-- `a` is existentially quantified
printSomeStuff :: (forall a. Show a => a -> String) -> IO ()
```

In the first case, `a` is instantiated at the use-site (where the function `head` is used/applied):

```hs
-- `a` is instantiated to `Int`
result = head @Int []
```

In the second case, `a` is instantiated at the definition-site (where the function `printSomeStuff` is defined):

```hs
printSomeStuff :: (forall a. Show a => a -> String) -> IO ()
printSomeStuff printFunction = do
  -- `a` is instantiated to `Int`
  putStrLn (printFunction @Int 123)
  -- `a` is instantiated to `Char`
  putStrLn (printFunction @Char 'c')
```

Reviewer 4: 👏

### Usage

Let's try to add `forall.` manually.
The syntax is very simple – at the beginning of function or instance declaration, before any constraints or arguments are used, use
the quantifier to introduce all type variables you want to use later:

```Haskell
func :: forall a b c m. a -> b -> m c -- OK, it compiles
```
Here we introduce four variables.
All type variables that you introduce in your definition should belong to one `forall` exactly, though they may not belong to the same one. (I will explain it further.)

```haskell
func :: forall a b c.   a -> b -> m c -- Error, type variable `m` is not introduced
```

We might want to introduce some constraints.
For example, we might want to specify that the `m` variable here is an instance of `Monad` typeclass.

> In order to add this constraint, we need to place it after all variables are declared:

Reviewer 2: > In order to add this constraint, we need to place it after all variables are declared

This seems like a repetition of what you said above:

> before any constraints or arguments are used, use
the quantifier to introduce all type variables

Perhaps this paragraph is not needed?

```Haskell
func :: forall a b c m. Monad m => a -> b -> m c
```
You might want to ask, what's so special about this syntax?
Nothing changed since we added the quantifier because it was already there!
However, there are a few things to notice here.

### Examples

#### Order of type variables

First of all, you can now change the order of type variables.
Let's imagine you want to define a function with 10 type variables (yeah, there are real-life examples of such functions), and you know that when you use it, you want to specify the last argument explicitly.

> To pass a type as a parameter to a function, we will use the `TypeApplication` extension.

Reviewer 4: This is a very expert sentence to write. You'll lose a lot of the audience, me included, here.
I might come back to the flow after I see the code below, but I also might as well stop reading.

 - [x] Explain what are _parameters_, 
 - [x] Show a distilled actual example you claim exists, 
 - [x] Explain why do you need to define `func` below.
 
It's a nice example and a nice section that has a chance to teach the readers a lot about more obscure techniques, but it needs to be done carefully, not by simply throwing some advanced intermediate stuff at beginner readers :)

veryLongFunction = ...

func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
```
Quite long, right?
This is because GHC will maintain the same order as when these variables first appear in the function definition.
However, by using explicit `forall.`, this can be simplified:

```Haskell
veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
veryLongFunction = ...

func = veryLongFunction @Integer ...
```

Since `j` is the first type variable in the declaration of `veryLongFunction` you can specify only it.
This is useful not only in such big examples but also when your function requires any of its arguments to be specified on usage.

### Support of other extensions

`ExplicitForAll` shines most when you need to enable other extensions – some of them just won't work without the `forall.` quantifier. :)
The most common one is `ScopedTypeVariables`, which allows you to pass a type variable from an outer function to the one defined in a where clause.

```haskell
example :: a -> [a] -> [a]
example x rest = double ++ rest
  where
    double :: a -> [a]
    double x = [x, x]
```

> This will cause an error without `ScopedTypeVariables`. But to use it, you need to add a `forall.` quantifier.

Reviewer 2: _Why_ does it cause an error?

Reviewer 4: +

Author: done

Reviewer 2: Done: https://github.com/serokell/blog-posts/pull/4/commits/aabbf1fb65912d7ee8dcfc2f2698b6f8a4f52719

```haskell
example :: forall a. a -> [a] -> [a]
```
You might also notice how it was useful with the `TypeApplication` extension.

> Other extensions that somehow benefit from the usage of `ExplicitForAll` are `LiberalTypeSynonyms`, `RankNTypes`, and many more.

Reviewer 4: Haskell beginners see RankNTypes and ScopedTypeVariables way more often than TypeApplication and LiberalTypeSynonyms, at least that's how I feel. There was also some extension which allows to write some inline signatures or something, I don't remember how it's called (maybe it's RankNTypes), I wonder if ExplictForall shines there too.

Reviewer 4: Read the post a little bit more and understood that I meant existential types lol lol

Reviewer 4: Actually, wait a second, isn't existential signature you show below a rank 2 type?! If so, I "yes", I would suggest throwing away `veryLongFunction` and explaining RankNTypes, but also, if "yes", you must mention it when presenting your `forall b. (Show b) => b`. If "no", then I **strongly** suggest throwing away `veryLongFunction` section anyway, explaining RankNTypes and then explaining the difference between rank 2 type and existential you show below.

Reviewer 2: > Actually, wait a second, isn't existential signature you show below a rank 2 type?

Yeah, in `func`, the rank of the type variable `b` is 2.

Reviewer 2: My terminology was a bit off: type vars don't have ranks, _types_ do.

So what I should have said is: `forall a c. a -> (forall b. (Show b) => b) -> c` is a rank-2 type, therefore it contains an existentially quantified type var (i.e., `b`).

Reviewer 2: > suggest (...) explaining RankNTypes

Since we're gonna be telling readers they need to enable `RankNTypes`, I agree, we should at least give a brief description of what rank-2 types are.

The two topics are _very_ intertwined.
In fact, if you start by explaining ranks first, then it becomes a lot easier to explain existentially quantification afterwards.

@PruStephan: I remember you mentioning you were having a hard time defining existential quantification, so this may actually work in your favour!

(BTW, the [GHC user guide](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/rank_polymorphism.html) is a good place to start)

Here's a suggestion: just as you're beginning to talk about existential quantification, you could begin by saying something like:

> In order to proceed further, we'll have to enable the `RankN` extension. This lets us write higher-rank types.
>
> By default, all types are rank-1. In a rank-1 type, a `forall` cannot appear on the left of an arrow (keep in mind that the `->` operator has precedence over `.`)
>
> (Examples of rank-1 types)
>
> With the `RankN` extension, we can write rank-2 types; that is, a type where a `forall` appears to the left of one arrow:
>
> (Use the `func` example from below)
> func :: (forall a. Show a => a -> IO ()) -> IO ()
>
> Because `forall a` appears to the left of one arrow, `a` is said to be existentially quantified.
> But what does this mean? (And THEN explain what existentially quantification means)

Reviewer 2: I thinking pointing out the general case may also be useful: if a type variable is introduced to the left of an even number of arrows (0, 2, 4, ...), then it's universally quantified. If it's introduced to the left of an odd number of arrows (1, 3, 5, ...), then it's existentially quantified.

Reviewer 2: > I would suggest throwing away `veryLongFunction`

@serokellcao  Why do you think it should be thrown away?

IMO, I think it's important to show that `forall` can be used to control the order of type applications.
I do think this specific example is very unrealistic, and the reader will have a hard time relating. If we want to convince the reader that this is useful, we should replace this example with a more realistic one, something that they might actually use in a project.

Author: Well, there is the same looking example from hoogle, which I placed a link for in text

Author: > Haskell beginners see RankNTypes and ScopedTypeVariables way more often than TypeApplication 

I am not sure if it is true, actually. I think that explaining what we use `TypeApplication` for is far easier, than `RankNTypes`

## Existential quantification
Besides universally quantified variables, Haskell also supports existential types.

> But in order to use them ... you still need to use the `forall.` keyword ¯\\\_(ツ)_/¯.

Reviewer 4: I'm rooting for responsible journalism. [There was `exists`, but it was deprecated](https://web.archive.org/web/20060304191310/https://hackage.haskell.org/trac/haskell-prime/wiki/ExistentialQuantification#Syntaxofexistentials). Your post would better be, I guess, the only modern post about existentials to answer the question "why". :)

Author: Actually, Beside this article, I have seen only that it probably was added in jhc compiler, but that is it. No other clues, that it was actually added to ghc, or if it was removed from plans, without even implementing it

The following construction is called an existential type in Haskell – when a type variable is introduced (using the `forall.` quantifier) inside the signature, not before it.
In contrast to the universal quantifier, the existential one means that the compiler (**callee**) is responsible for deducing what a type variable should mean.
This also means that the caller can't specify this type variable.

![When you use existential types](meme.jpg)

### Usage

We will start with functions:

```haskell

> func :: forall a c. a -> (forall b. (Show b) => b) -> c

Reviewer 2: Also, I don't think this function is a good example, because it cannot possibly be called.

There's no way you can pass a `(forall b. Show b => b)` to `func`.

Let's take a step back: because `b` is existentially quantified, `func`'s definitinion is gonna be responsible for instantiating `b` to some type.

So, at the _use-site_, when you're _using_ func, you have to pass in a function of type `(forall b Show b => b)`. In other words, you have to pass in a function that creates a variable of type `b` no matter what `b` is. That can't be done.

The only way to call this function is to pass in `undefined`.

Reviewer 2: The function you replaced it with (`func :: forall a c. a -> (forall b. (Show b) => b -> IO ()) -> IO c`) can be called, but it _can't be defined_ (without using `undefined`).

Author: Well, I would say, that currently this function is some sort of `log` from previous example. What it should do here is a different question. Probably in real, it would be nice to have some constraint CanCast a b for example. But I guess it becomes to complicated for this example

Reviewer 2: This was changed in a recent commit, marking this as resolved.

```
There are a few things to mention here.

* In the type signature, `a` and `c` are universally quantified variables, and `b` is existentially quantified. This is what I was talking about when I said that type variables may belong to different quantifiers.
* The same variable can't be quantified more than once.
* You can see that constraints for `b` are added in the middle of the declaration of the function type (because `b` was introduced in the middle of the function).

> * Now you can't manually choose type for `b` – the compiler must deduce the correct variable you want to use by itself.

Reviewer 2: > the compiler must deduce the correct variable you want to use by itself

As per my previous comments: this is not true either.

Here's a real-life example of using existential types in functions.
Imagine a function that prints some logs while calculating the result.
You may want to write logs into a console or into some file.
In order to do it, you pass a logging function as an argument.

```haskell
func :: _ -> IO ()
func log = do
  username <- getUsername
  log username

  userCount <- getUserCount
  log userCount

getUsername :: IO String
getUsername = undefined

getUserCount :: IO Int
getUserCount = undefined
```

What type should the log function have?
Since it somehow logs both `String` and `Int`:

```haskell
log :: Show a => a -> IO ()
```
However, if we add this signature to `func` and try to call it with different arguments:

```haskell
func :: Show a => (a -> IO ()) -> IO ()
func log = ...

main :: IO ()
main = do
  -- log to the console
  func print

  -- log to a logfile
  func (appendFile "logfile.log" . show)
```
We will see some nasty compiler errors:
```

> Ambiguous type variable ‘a1’ arising from a use of ‘func’

Reviewer 2: The way this is phrased, it appears that the problem with the code snippet above is in `main`.

But it's not - the problem is in `func`. `func`'s definition is trying to instantiate `a=String`, but `a` is universally quantified, and that's why it can't be compiled. `func`'s definition cannot instantiate `a`, only the caller of `func` can.

![image](https://user-images.githubusercontent.com/718166/152653881-72475ac8-0463-4d6e-8be2-2d073588ce1e.png)

Yes, there's a "Ambiguous type variable" error in `main`, but that's a red herring. The real cause of the issue is in `func`.

Reviewer 4: 👏 

@dcastro your comments are the reason why expert reviews are as important, if not more important as "beginners' mind" reviews. Amazing, stellar work! 

I want to say that I don't want to diminish @PruStephan's work too. Writing blogs are hard and for the first attempt, this is a very very good start!

Reviewer 2: Much appreciated! :smile: :bow: 

> Writing blogs are hard and for the first attempt, this is a very very good start!

Indeed! I want to echo your words.

Writing blog posts is also a great learning & discovery process.
You don't necessarily have to know everything about a topic _before_ putting the pen to the paper.

@PruStephan I can tell you've learned a lot since you began writing this! Which can only mean you're doing a good job :smile:

Author: Thanks. That is actually was the reason I am writing this post, to learn something new. I think I after this post, I will continue with writing

```

This is because the compiler can't deduce the type of `a` from the given constraints.
If we use `forall.` explicitly as we did before, it wants help, because in this case, the caller is the one who should specify the type of the log function. But we can't specify it because it is different for different logging operations.
The only solution here is to add constraints and `forall.` quantifier to the log function signature:

```haskell
{-# LANGUAGE RankNTypes #-}

func :: (forall a. Show a => a -> IO ()) -> IO ()
```

> Now we put all responsibility for deducing types on the shoulders of the compiler, and it will deduce them the way it wants.

Reviewer 4: Show that it works and say a couple of words of what did the compiler infer under the hood.

Author: This sentence was removed from the article

#### Hiding type variables

```Haskell
data MyData a = forall b. MkMyData b a
-- You need to enable either `ExistentialQuantification` or `GADTs` extension to use this syntax
```

The first thing you might have noticed is that we don't have a type variable for `b` on the left side of the expression.
However, we still can use it on the right side.
This is because `b` is existentially quantified.
The compiler will deduce the type of the variable itself, and you can't specify it.

```haskell
func :: MyData Integer -> IO ()

> func (MkMyData (x :: Double) y) = print x -- compiler error - can't match type of x with Double

Reviewer 4: Ok, but how do I use it meaningfully?

Also, an interesting take: to me it looks like co-phantom data type. :)

Author: > How do I use it meaningfully

You are talking about `MyData` type?  Well, without constraints (as I am showing below) there is literally no way to use  `x` in function

Reviewer 2: Then maybe consider replacing `MyData` with something that can be meaningfully used.

Remember: you're trying to convince the reader that these things _are useful_ and therefore they should pay close attention to your explanation.

When you use examples that are not meaningful or are too farfetched/contrived, the reader may begin scratching their head and stop reading.

Author: I moved this example to SomeException section, with explaining that you need to use special function to extract exception

```

> #### Heterogeneous lists

Reviewer 4: THIS EXAMPLE IS THE HOT STUFF OF THIS POST.

People get crushed trying to implement heterogenous structures in Haskell and your post really excels making heterogenous structures LOGICAL in Haskell. BRAVO!

Author: Thanks)

However, adding constraints to the `a` variable may improve the situation:

```haskell
data Elem = forall a. (Show a) => Elem a

multitypedList :: [Elem]
multitypedList = [Elem "a", Elem 1, Elem (Just 5)]

printElem :: Elem -> IO ()
printElem (Elem x) = print x

main :: IO ()
main = fmap printElem multitypedList

λ> "a"
λ> 1
λ> Just 5
```

Note that since type variable `a` is hidden with a quantifier, you can't add constraints to it besides those added to the constructor.

```haskell
showFunc :: Show ??? => [Elem] -> String -- There is no type variable, that you can add constraint to
showFunc (x:xs) => (show x) ++ (showFunc xs)
```

We hide exception type in the same way in the well-known `SomeException` wrapper:

```haskell
data SomeException = forall a. Exception a => SomeException a
```

It allows you to catch exceptions of any type.
As long, as you hadn't specified its type, (and it is not obvious to the compiler, from the handler for example), all exceptions will be catch, wrapped in `SomeException` datatype.
After that, you can use `fromException` function, to check, that you caught something you wanted to catch.

> ## Summary

Reviewer 4: Good post. Not much left to fix.
Pay special attention to Diogo's feedback, dig the history of GHC trac ticket #26 and explain why exists was replaced by just forall. Do some minor tweaks as per my review too.

Amazing job, @PruStephan !

Without `forall`, you would write something like this:

```Haskell
veryLongFunction :: a -> b -> c -> d -> e -> f -> g -> h -> i -> j
veryLongFunction = ...

func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
```
Quite long, right?
However, by using explicit `forall`, this can be simplified:

```Haskell
veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
veryLongFunction = ...

func = veryLongFunction @Integer ...
```

Since `j` is the first type variable in the declaration of `veryLongFunction` you can specify only it.
This is useful not only in such big examples but also when your function requires any of its arguments to be specified on usage.

### Support of other extensions

`ExplicitForAll` shines most when you need to enable other extensions – some of them just won't work without the `forall` quantifier. :)
The most common one is `ScopedTypeVariables`, which allows you to pass a type variable from an outer function to the one defined in a where clause.

```haskell
example :: a -> [a] -> [a]
example x rest = pair ++ rest
  where
    pair :: [a]
    pair = [x, x]
```

> This will cause an error without `ScopedTypeVariables`, because compiler thinks that `a` in `example`'s signature and `a` in `double`'s are two different types.

Reviewer 3: >because compiler thinks that `a` in `example`'s signature and `a` in `double`'s are two different types.

AFAIU, the error happens because the compiler takes `pair` function to be a polymorphic one, which should be able to return a list of values of *any* type. But its implementation `[x, x]` can only be a single concrete type, as dictated by the concrete type variable from the `example` function. 

So I think this can be easily corrected by saying "two different type variables" instead of "two different types".

Reviewer 2: @sras I expanded on the explanation of `ScopedTypeVariables` here: https://github.com/serokell/blog-posts/pull/4/commits/aabbf1fb65912d7ee8dcfc2f2698b6f8a4f52719

```
There are a few things to mention here.

* In the type signature, `a` and `c` are universally quantified variables, and `b` is existentially quantified.
* The same variable can't be quantified more than once.
* You can see that constraints for `b` are declared _after_ `b` has been introduced with the `forall` quantifier
* You can choose type for `b` only inside `func` definition, and you can't do it when you call `func` (we will discuss it further below).

Here's a real-life example of using existential types in functions.
Imagine a function that prints some logs while calculating the result.
You may want to write logs into a console or into some file.
In order to do it, you pass a logging function as an argument.

```haskell
func :: _ -> IO ()
func log = do
  username <- getUsername
  log username

  userCount <- getUserCount
  log userCount

getUsername :: IO String
getUsername = undefined

getUserCount :: IO Int
getUserCount = undefined
```

What type should the log function have?
Since it somehow logs both `String` and `Int`:

```haskell
log :: Show a => a -> IO ()
```
However, if we add this signature to `func` and try to call it with different arguments:

```haskell

> func :: Show a => (a -> IO ()) -> IO ()

Reviewer 3: In the previous sentance you say "if we add this signature to `func`". But if we do that, then the resulting function should be 
```
 func :: (Show a => (a -> IO ())) -> IO ()
```
since `Show a` constraint come from the `log` function.

With this change we can instantiate `a` when we call log.

```haskell
func :: (forall a. Show a => a -> IO ()) -> IO ()
func log = do
  username <- getUsername
  log @String username

  userCount <- getUserCount
  log @Int userCount
```

Note, that this type applications is unnecessary.
Compiler is smart enough to choose appropriate types.

#### Heterogeneous lists

Now we understand that existential types allow you to introduce type variables without the necessity to specify them.
Let's take a look how this is done in datatypes.
You can introduce existentially quantified type variable in the datatype, by using `forall` on the left side of the constructor name. Existential types in datatypes allow you to have field in the datatype, without necessity to specify its type.

```haskell
data Elem = forall a. Elem a

multitypedList :: [Elem]
multitypedList = [Elem "a", Elem 1, Elem (Just 5)]
```

This is a heterogeneous list that contains values of different types.
You can't do that with an universally quantified list because every element in that list would need to have the same type.
But, there are not many things you can do with elements of this list.

> Since the type of elements is unknown at the use-site, you can't address it, nor can you pass it to functions that expect values of concrete type.

Reviewer 3: > Since the type of elements is unknown at the use-site, you can't address it, nor can you pass it to functions that expect values of concrete type.

Maybe here or below, we can show the ability of `Typeable` typeclass to bring back concrete type into scope, enabling these values to be passed to functions that expect concrete types.

Author: done

However, adding constraints to the `a` variable may improve the situation:

```haskell
-- You need to enable either `ExistentialQuantification` or `GADTs` extension to use this syntax
data Elem = forall a. (Show a) => Elem a

multitypedList :: [Elem]
multitypedList = [Elem "a", Elem 1, Elem (Just 5)]

-- We can use 'print' here because inner types of 'Elem' has 'Show' constraint
printElem :: Elem -> IO ()
printElem (Elem x) = print x

main :: IO ()
main = fmap printElem multitypedList

λ> "a"
λ> 1
λ> Just 5
```

> Note that since type variable `a` is hidden with a quantifier, you can't add constraints to it besides those added to the constructor.

Reviewer 3: This statement and the example below feels a bit weird. Isn't it obvious that you cannot add constraints to a type variable that is not in scope? It is also not clear why we are trying to do that here.

Author: Well, it is, but why not to mention it? I mean, It is better to explicitly state that it is impossible, for those who still hope, that there is a way...

Reviewer 3: >but why not to mention it?

For sake of brevity I guess. Basically in articles like these, it does not often help to put as much detail as possible, without carefully arranging it. So unless there is a really good reason to think that the reader will naturally wonder if they can add a constraint to a hidden type variable, I think it is better to omit it. Not a big deal though, I ll leave it upto you.

Reviewer: I think this is a reasonable thing to leave, it fits quite well.

```haskell
showFunc :: Show ??? => [Elem] -> String -- There is no type variable, that you can add constraint to
showFunc (x:xs) => (show x) ++ (showFunc xs)
```

Other useful example of hiding type variables is `SomeException` wrapper

```haskell
data SomeException = forall a. Exception a => SomeException a
```

It allows you to catch exceptions of any type.
As long, as you hadn't specified its type, (and it is not obvious to the compiler, from the handler for example), all exceptions will be caught, wrapped in `SomeException` datatype.

However, again, there is a limitation.
We know nothing about inner exception.
Yes, we can for example print it, but sometimes we need more information.
We can't use type inheritance to describe, what type we want to get

```haskell
-- Something like that won't work.
-- we know to little about type inside,
-- to be sure that it is 'MyException'
myFromException :: SomeException -> MyException
myFromException (SomeException (ex :: MyException)) = ex
```

> Luckily for users, module `Control.Exception` provides special function `fromException`, the purpose of which is to extract inner variable and cast to specified type. How it is done on the inside, is a topic of a different article.

Reviewer 3: >inner variable

Better say 'inner value', or 'wrapped value'.

# Existential and Universal Quantification in Haskell

In this article I give a brief introduction into one of the most common Haskell extensions – `ExplicitForAll` – and syntax constructions that this extension brings to language.
It is used in about all big haskell projects, so it is very nice to either become acquainted with it or to refresh it in your memory if you already know about it.

> Stay tuned if you want to know how to create lists of differently typed variables, how to add complex constraints, and more!

Reviewer 4: Nice

## Introduction

> By default, quantification in Haskell is implicit.

Reviewer 4: Beginner will be confused as heck here. Explain what is quantification, illustrate explicit vs implicit.

Reviewer: Yes, I agree. And also what is a quantifier in general. Will provide examples soon.

Without the `ExplicitForAll` extension, you can't write quantifiers. Moreover, all variables are universally quantified, and existential quantification is off.

> The key point of both universal and existential quantification is choosing where, we are going to instantiate a type variable - where it is defined, or where it is used.

Reviewer 4: As you can see, the notion of instantiation can be confusing too. I would spare a couple of words to explain that instantiation is turning a variable into a concrete type. It's rather different, by the way, from what Java people think instantiation is, which -- in their case -- is all about bringing variables to scope.

https://stackoverflow.com/questions/23706331/terminology-what-does-it-mean-to-instantiate-type-variables-in-prof-giesls

Reviewer 2: Agreed, very good point!

Tweag also give a similar definition in [a blog post](https://www.tweag.io/blog/2020-04-02-lazy-eager-instantiation/): 

> When inferring the type for `myConst`, the compiler _instantiates_ the type of const, meaning that the type variables `a` and `b` get replaced by (potentially yet unknown) types.

Reviewer 2: Done here: https://github.com/serokell/blog-posts/pull/4/commits/63b2e18f537a32f858d243ace6ccb563710a4095

## Universal quantification and `ExplicitForAll`

By default, all type variables are considered universally quantified.

> This means that every time you use a function like `id`, you can _instantiate_ the universally quantified type variable `a` to (i.e., replace it with) any type.

Reviewer 4: Related to my comment up there: why won't we explain what "instantiation" means at the first occurrence?

Reviewer 2: Done here: https://github.com/serokell/blog-posts/pull/4/commits/63b2e18f537a32f858d243ace6ccb563710a4095

### Usage

> Let's try to add `forall` manually.

Reviewer 4: `forall` is not defined yet, it will help if you show `forall` in the intro and talk about what it is just a bit. This article assumes a lot of Haskell competence.

Reviewer: Agree, if you explain what is a quanitifier correctly, you will be able to say that we are adding a quantifier and that's all.

The syntax is very simple – at the beginning of function or instance declaration, before any constraints or arguments are used, use
the quantifier to introduce all type variables you want to use later:

```Haskell
func :: forall a b c m. a -> b -> m c -- OK, it compiles
```
Here we introduce four variables.
All type variables that you introduce in your definition should belong to one `forall` exactly, though they may not belong to the same one. (I will explain it further.)

```haskell
func :: forall a b c.   a -> b -> m c -- Error, type variable `m` is not introduced
```

Let's add some constraints.
For example, we might want to specify the `m` variable here as an instance of `Monad` typeclass.

```Haskell
func :: forall a b c m. Monad m => a -> b -> m c
```
You might want to ask, what's so special about this syntax?

> Nothing changed since we added the quantifier because it was already there!

Reviewer 4: It's confusing. If it's already there, why didn't `forall a b c . a -> b -> m c` work? `m` should be in the implicit, no?

Author: I don't get your point. We can't left `m` be quantified implicitly, or we get an error

# Existential and Universal Quantification in Haskell

This article gives a brief introduction into one of the most common Haskell extensions – `ExplicitForAll` – and syntax constructions that this extension brings to language.
It is used in about all big haskell projects, so it is very nice to either become acquainted with it or to refresh it in your memory if you already know about it.

Stay tuned if you want to know how to create lists of differently typed variables, how to add complex constraints, and more!

## Introduction

> By default quantification in Haskell is implicit. Without special extension - `ExplicitForAll`, quantifiers will stay hidden. However, enabling it allows us to see the true functions types, like for example, the type of `id`.

Reviewer 2: I wouldn't say the extension is _special_ (it's not different from any other extension), so how about rephrasing this a bit?

```suggestion
By default, quantification in Haskell is implicit. Without the `ExplicitForAll` extension, quantifiers will stay hidden.
However, enabling it allows us to see the true functions types, like for example, the type of `id`.
```

Author: done

> By default quantification in Haskell is implicit. Without special extension - `ExplicitForAll`, quantifiers will stay hidden. However, enabling it allows us to see the true functions types, like for example, the type of `id`.

Reviewer 2: > enabling it allows us to see the true functions types

@sras's previous comment (https://github.com/serokell/blog-posts/pull/4#discussion_r782011902) applies here too.

The extension does not let you see the quantifier, it lets you write it explicitly

Author: Seems to be done

```haskell
id :: forall a. a -> a
id x = x
```

This quantifier comes from predicate logic.
From there we also know about existence of the `exists` quantifier.

> While, there were some attempts to use it [in the past](https://web.archive.org/web/20060304191310/https://hackage.haskell.org/trac/haskell-prime/wiki/ExistentialQuantification#Syntaxofexistentials), it was removed from the compiler, in favor of using only `forall` one.

Reviewer 2: Extra commas here:

* > While,
* > quantification is choosing where, we are going
* > means is chosen, when you call

Author: Some day I will stop inject Russian grammar rules into my English writing. You know, Russian grammar on commas is a special type of hell

Author: done

> While, there were some attempts to use it [in the past](https://web.archive.org/web/20060304191310/https://hackage.haskell.org/trac/haskell-prime/wiki/ExistentialQuantification#Syntaxofexistentials), it was removed from the compiler, in favor of using only `forall` one.

Reviewer 2: > it was removed from the compiler

I don't think it was ever added in the first place...

> While, there were some attempts to use it [in the past](https://web.archive.org/web/20060304191310/https://hackage.haskell.org/trac/haskell-prime/wiki/ExistentialQuantification#Syntaxofexistentials), it was removed from the compiler, in favor of using only `forall` one.

Reviewer 2: > From there we also know about existence of the `exists` quantifier.
> (...) it was removed from the compiler, in favor of using only `forall` one.

The way this is phrased, a reader might interpret it to mean "existential quantification is not possible in Haskell".

Let's try to use a more clear language. You wanted to say that existential quantification is possible even though there is no `exists` **keyword** in Haskell. So, instead, we repurpose the `forall` keyword to implement existential quantification.

----

Additionally, this is a technical detail, and as such, I don't think the "Introduction" section is the right place to go into this much detail.

Perhaps this detail should be moved to the "Existential quantification" section.

Author: done

The key point of both universal and existential quantification is choosing where, we are going to instantiate a type variable - where it is defined, or where it is used.

## Universal quantification and `ExplicitForAll`

By default, all type variables are considered universally quantified.
It means that what type variable means is chosen, when you call a function.

> For example, every time you call `id` in other functions, type variable `a` is specified with concrete type.

Reviewer 2: > It means that **what type variable means is chosen**, when you call a function.
> For example, every time you call `id` in other functions, type variable `a` **is specified with** concrete type.

The expression you're looking for is "instantiation" (of a type variable). Consider rephrasing this paragraph like this:

```suggestion
This means that every time you use a function like `id`, you can _instantiate_ the universally quantified type variable `a` to (i.e., replace it with) any type.
```

Author: done

### Usage

Let's try to add `forall` manually.
The syntax is very simple – at the beginning of function or instance declaration, before any constraints or arguments are used, use
the quantifier to introduce all type variables you want to use later:

```Haskell
func :: forall a b c m. a -> b -> m c -- OK, it compiles
```
Here we introduce four variables.
All type variables that you introduce in your definition should belong to one `forall` exactly, though they may not belong to the same one. (I will explain it further.)

```haskell
func :: forall a b c.   a -> b -> m c -- Error, type variable `m` is not introduced
```

> Lets add some constraints.

Reviewer 2: ```suggestion
Let's add some constraints.
```

Author: done

For example, we might want to specify the `m` variable here as an instance of `Monad` typeclass.

```Haskell
func :: forall a b c m. Monad m => a -> b -> m c
```
You might want to ask, what's so special about this syntax?
Nothing changed since we added the quantifier because it was already there!
However, it allows us to use new tricks.

### Examples

#### Order of type variables

First of all, you can now change the order of type variables.
Let's imagine you want to define a function with 10 type variables (yeah, there are real-life [examples](https://hackage.haskell.org/package/leancheck-0.9.10/docs/Test-LeanCheck-Utils-TypeBinding.html#v:-45--62--62--62--62--62--62--62--62--62--62--62--62-:) of such functions), and you know that when you use it, you want to specify type of the last argument explicitly.
To pass a type as a parameter to a function, we will use the `TypeApplication` extension.

> Few notes about passing types to functions. Using `TypeApplication` extension you can pass types as function arguments, using `@` sign (If specifying concrete type is unnecessary you can use type wildcard `@_`).

Reviewer 2: > To pass a type as a parameter to a function, we will use the `TypeApplication` extension

> Few notes about passing types to functions

> Using `TypeApplication` extension you can pass types as function arguments

You're repeating yourself.

Author: done

Without `forall`, you would write something like this:

```Haskell
veryLongFunction :: a -> b -> c -> d -> e -> f -> g -> h -> i -> j
veryLongFunction = ...

func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
```
Quite long, right?
However, by using explicit `forall`, this can be simplified:

```Haskell
veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
veryLongFunction = ...

func = veryLongFunction @Integer ...
```

Since `j` is the first type variable in the declaration of `veryLongFunction` you can specify only it.
This is useful not only in such big examples but also when your function requires any of its arguments to be specified on usage.

### Support of other extensions

`ExplicitForAll` shines most when you need to enable other extensions – some of them just won't work without the `forall` quantifier. :)
The most common one is `ScopedTypeVariables`, which allows you to pass a type variable from an outer function to the one defined in a where clause.

```haskell
example :: a -> [a] -> [a]
example x rest = pair ++ rest
  where
    pair :: a -> [a]

>     pair x = [x, x]

Reviewer 2: This example doesn't compile because `pair` takes an argument of type `a` and you've given it none.

Even after changing `pair` to `pair x`, the example now compiles **without** the `ScopedTypeVariables` extension.

Author: done

```
This will cause an error without `ScopedTypeVariables`, because compiler thinks that `a` in `example`'s signature and `a` in `double`'s are two different types.
To fix it, you also need to use extension and add a `forall` quantifier, as extension won't work without it.

```haskell
example :: forall a. a -> [a] -> [a]
```
You might also notice how it was useful with the `TypeApplication` extension.
Other extensions that somehow benefit from the usage of `ExplicitForAll` are `LiberalTypeSynonyms`, `RankNTypes`, and many more.

## Existential quantification
As I said earlier, besides universally quantified variables, Haskell also supports existential types.
Remember, that we still use the `forall` keyword ¯\\\_(ツ)_/¯.

> Actually, it is implemented this way, because in terms of first-order predicate logic these two constructions are equivalent: `(exists x. p x) -> q` and `forall x. (p x -> q)`.

Reviewer 2: Is mentioning this equivalence and predicate logic necessary? I fear it might distract the reader from the point of the blog post.

Connecting the dots between the "predicate logic encoding" and the "haskell encoding" of existential quantification could be useful _if_ we expected most readers to be familiar with predicate logic, but I don't think that's a reasonable expectation...

Author: Well, it is good to mention, why something called existential instead is written using `forall`. Even if the reader doesn't know anything about FOPL they will still notice, that there is some clever explanation to it

Reviewer: I think it is good to leave it there since it gives some proof that what we are doing with forall has connection with existentials. It's nice to spell it out.

If you want a theoretical proof of this statement, you can check this [tread](https://stackoverflow.com/questions/10753073/whats-the-theoretical-basis-for-existential-types?rq=1).

### Usage

In order to proceed further, we'll have to enable the `RankNTypes` extension. This lets us write higher-rank types.

By default, all types are rank-1. In a rank-1 type, a `forall` cannot appear on the left of an arrow (keep in mind that the `->` operator has precedence over `.`)

> For example, this function have rank-1 types:

Reviewer 2: ```suggestion
For example, this function has a rank-1 type:
```

Author: done

```haskell

> f :: forall a b. a -> b -> a

Reviewer 2: ```suggestion
f :: forall a b. a -> b -> a

-- Or, to make the precedence explicit:
f :: forall a b. (a -> b -> a)
```

Author: done

```

With the `RankNTypes` extension, we can write rank-2 types - types where a `forall` appears to the left of one arrow:

```haskell
f :: (forall a. Show a => a -> IO ()) -> IO ()
```

> Because forall a appears to the left of one arrow, `a` is said to be existentially quantified.

Reviewer 2: ```suggestion
Because `forall a` appears to the left of one arrow, `a` is said to be existentially quantified.
```

Author: done

It is also true for every `forall` that is introduced to the left of an odd number of functional arrows.
But what does this mean?

In contrast to the universal quantifier, the existential one means that the function is itself responsible for instantiating what a type variable should mean. So the result type of existentially quantified variable will become known when we address it at the 'definition-site' (i.e., in the function body).
This also means that no one can specify its type outside of function.

Let's take a closer look on practical examples.

We will start with functions:

```haskell
func :: forall a c. a -> (forall b. (Show b) => b -> IO ()) -> IO c
```
There are a few things to mention here.

* In the type signature, `a` and `c` are universally quantified variables, and `b` is existentially quantified.
* The same variable can't be quantified more than once.

> * You can see that constraints for `b` are added in the middle of the declaration of the function type (because `b` was introduced in the middle of the function).

Reviewer 2: "in the middle of the declaration" is a bit vague and ambiguous... how about "constraints for `b` are declared _after_ `b` has been introduced with the `forall` quantifier" (with emphasys on `_after_`)

Author: done

* You can choose type for `b` only inside `func` definition, and you can't do it when you call `func` (we will discuss it further below).

Here's a real-life example of using existential types in functions.
Imagine a function that prints some logs while calculating the result.
You may want to write logs into a console or into some file.
In order to do it, you pass a logging function as an argument.

```haskell
func :: _ -> IO ()
func log = do
  username <- getUsername
  log username

  userCount <- getUserCount
  log userCount

getUsername :: IO String
getUsername = undefined

getUserCount :: IO Int
getUserCount = undefined
```

What type should the log function have?
Since it somehow logs both `String` and `Int`:

```haskell
log :: Show a => a -> IO ()
```
However, if we add this signature to `func` and try to call it with different arguments:

```haskell
func :: Show a => (a -> IO ()) -> IO ()
func log = ...

main :: IO ()
main = do
  -- log to the console
  func print

  -- log to a logfile
  func (appendFile "logfile.log" . show)
```

> We will see some nasty compiler errors:

Reviewer 2: On which line?

Author: done

```
Couldn't match expected type `a` with actual type `String`...
```
It is happening, because inside of `func`, `a` is not known yet. I will be known, only when you call `func` somewhere.

> The only solution here is to add constraints and `forall` quantifier to the log function signature:

Reviewer 2: Let's try to use a unified language throughout the blog post.

Instead of saying "the type is not known yet" or "the type will be known", let's keep using the "instantiate" expression we've been using before.

By using a consistent language, the reader can build an intuition for what these terms mean.

```suggestion
We're getting this error because:
1. `func`'s type variable `a` is universally quantified, therefore it can only be instantiated at the use-site.
2. We're trying to instantiate it at `func`'s definition-site (`log username` is trying to instantiate the type variable `a` with `String` and `log userCount` is trying to instantiate the type variable `b` with `Int`).

Because we want `a` to be instantiated at `func`'s definition-site (rather than its use-site), we need `a` to be existentially quantified instead.
```

Author: done

```haskell
{-# LANGUAGE RankNTypes #-}

func :: (forall a. Show a => a -> IO ()) -> IO ()
```
Now the compiler knows, that whatever function we pass to `func`, type variable `a` should be instantiated here, not outside.
After this change, we can specify.

With this change we can instantiate `a` when we call log.

```haskell
func :: (forall a. Show a => a -> IO ()) -> IO ()
func log = do
  username <- getUsername
  log @String username

  userCount <- getUserCount
  log @Int userCount
```

Note, that this type applications is unnecessary.
Compiler is smart enough to choose appropriate types.

#### Hiding type variables

Let's turn to datatypes now.

> You can introduce existentially quantified type variable in the datatype, by using `forall` on the left side of equality sign.

Reviewer 2: "You can introduce... by..." > "One way of introducing... is by..."

There's actually _another_ way of doing the exact same thing:

```hs
data MyData a where
  MkMyData :: forall a b. b -> a -> MyData a
```

Author: Removed MyData part

Reviewer 2: You moved this part, but my comment still applies

Reviewer: I think "You can introduce" and "One way of introducing" are more or less equivalent.

> You can introduce existentially quantified type variable in the datatype, by using `forall` on the left side of equality sign.

Reviewer 2: "on the left side of equality sign" > "on the left side of the constructor name"

Author: done

```haskell
data MyData a = forall b. MkMyData b a
-- You need to enable either `ExistentialQuantification` or `GADTs` extension to use this syntax
```

The first thing you might have noticed is that we don't have a type variable for `b` on the left side of the expression.
However, we still can use it on the right side.
This is because `b` is existentially quantified.

> The compiler will instantiate the type of the variable itself, and you can't specify it.

Reviewer 2: Hmm this doesn't sound right... You can absolutely instantiate `b` when you're constructing a `MyData`:

```
md :: MyData Int
md = MkMyData @Int @String "" 1
```

Author: Removed MyData part

```haskell
func :: MyData Integer -> IO ()
func (MkMyData (x :: Double) y) = print x -- compiler error - can't match type of x with Double
```

#### Heterogeneous lists

Now we understand that existential types allow you to introduce type variables without the necessity to specify them.

This might be useful when building a multi-typed list:

```haskell
data Elem = forall a. Elem a

multitypedList :: [Elem]
multitypedList = [Elem "a", Elem 1, Elem (Just 5)]
```
This is a heterogeneous list that contains values of different types.

> You can't do that with an universally quantified list because every element in that list would need to have the same type.w

Reviewer 2: ```suggestion
You can't do that with an universally quantified list because every element in that list would need to have the same type.
```

Author: done

## Existential quantification

As I said earlier, besides universally quantified variables, Haskell also supports existential types.
But they also use the `forall` keyword.

It can be implemented this way because these two constructions – `(exists x. p x) -> q` and `forall x. (p x -> q)` – are equivalent in terms of first-order predicate logic.
For a theoretical proof of this statement, you can check this [thread](https://stackoverflow.com/questions/10753073/whats-the-theoretical-basis-for-existential-types?rq=1).

### Usage

In order to proceed further, we'll have to enable the `RankNTypes` extension. This lets us write higher-rank types.

<hr>

**Higher-rank types**

By default, all types are rank-1. In a rank-1 type, a `forall` cannot appear on the left of an arrow (keep in mind that the `->` operator has precedence over `.`)

For example, this function has a rank-1 type:

```haskell
f :: forall a b. a -> b -> a

-- Or, to make the precedence explicit:
f :: forall a b. (a -> b -> a)
```

With the `RankNTypes` extension, we can write rank-2 types – types where `forall` appears to the left of one arrow:

```haskell
f :: (forall a. Show a => a -> IO ()) -> IO ()
```

</hr>

> Because `forall a` appears to the left of one arrow, `a` is said to be existentially quantified.

Reviewer: Which is that arrow? 

This description confuses me, personally.

It is also true for every `forall` that is introduced to the left of an odd number of functional arrows.
But what does this mean?

In contrast to the universal quantifier, the existential one means that the function is itself responsible for instantiating what a type variable should mean. So the result type of existentially quantified variable will become known when we address it at the 'definition-site' (i.e., in the function body).
This also means that no one can specify its type outside of function.

Let's take a closer look on practical examples.

> We will start with functions:

Reviewer: I believe this is one function?

```haskell
func :: forall a c. a -> (forall b. (Show b) => b -> IO ()) -> IO c
```
There are a few things to mention here.

* In the type signature, `a` and `c` are universally quantified variables, and `b` is existentially quantified.
* The same variable can't be quantified more than once.
* You can see that constraints for `b` are declared _after_ `b` has been introduced with the `forall` quantifier
* You can choose type for `b` only inside `func` definition, and you can't do it when you call `func` (we will discuss it further below).

> Here's a real-life example of using existential types in functions.

Reviewer: I would start with an example.

Because we want `a` to be instantiated at `func`'s definition-site (rather than its use-site), we need `a` to be existentially quantified instead.

```haskell
{-# LANGUAGE RankNTypes #-}

func :: (forall a. Show a => a -> IO ()) -> IO ()
```
Now the compiler knows, that whatever function we pass to `func`, type variable `a` should be instantiated here, not outside.
With this change we can instantiate `a` when we call log.

```haskell
func :: (forall a. Show a => a -> IO ()) -> IO ()
func log = do
  username <- getUsername
  log @String username

  userCount <- getUserCount
  log @Int userCount
```

Note, that this type applications is unnecessary.
Compiler is smart enough to choose appropriate types.

#### Heterogeneous lists

Now we understand that existential types allow you to introduce type variables without the necessity to specify them.
Let's take a look how this is done in datatypes.
You can introduce existentially quantified type variable in the datatype, by using `forall` on the left side of the constructor name. Existential types in datatypes allow you to have field in the datatype, without necessity to specify its type.

```haskell

> data Elem = forall a. Elem a

Reviewer: Is it just me or it would be much easier to introduce the concept of existential quantification by comparing a universally quantified and existentially quantified list?

Author: Not sure .... There is no such thing as existentially quantified list. We could only have a regular list, where each element contains existentially quantified variable. And in order to explain how it works, we need to explain existentially quantified datatype first

Reviewer: Yeah, I mean a list where each element contains an existentially quantified variable. 

What I mean is that it is much easier to explain what existential quantification in Haskell is through an example. 

There are two good examples that you have: one with logging and the second one with list. 

So it would be good to start with either of them when exploring existential quantification. 

What I wanted to suggest is to focus on this list thing, since it feels simpler and then build up to the logging example (it is more useful) and the complete definition (it is a good way to structure what it _actually means).

Reviewer: Even more so because the hetero-list doesn't require RankNTypes, it seems? 🤔

`Data.Typeable` module's functionality can help to circumvent this restriction. It allows you to cast one type (even unknown) to another one.

Adding constraints to the `a` variable may also improve the situation:

```haskell
-- You need to enable either `ExistentialQuantification` or `GADTs` extension to use this syntax
data Elem = forall a. (Show a) => Elem a

multitypedList :: [Elem]
multitypedList = [Elem "a", Elem 1, Elem (Just 5)]

-- We can use 'print' here because inner types of 'Elem' has 'Show' constraint
printElem :: Elem -> IO ()
printElem (Elem x) = print x

main :: IO ()
main = fmap printElem multitypedList

λ> "a"
λ> 1
λ> Just 5
```
Note that since type variable `a` is hidden with a quantifier, you can't add constraints to it besides those added to the constructor.

```haskell
showFunc :: Show ??? => [Elem] -> String -- There is no type variable, that you can add constraint to
showFunc (x:xs) => (show x) ++ (showFunc xs)
```

Other useful example of hiding type variables is `SomeException` wrapper

```haskell
data SomeException = forall a. Exception a => SomeException a
```

It allows you to catch exceptions of any type.

> As long, as you hadn't specified its type, (and it is not obvious to the compiler, from the handler for example), all exceptions will be caught, wrapped in `SomeException` datatype.

Reviewer: I don't understand this sentence, there is something wrong with the way it is written, it is rather unclear.

Here and in the future examples, we will use the `TypeApplication` extension, which allows you to pass types as function arguments via the `@` sign. (If specifying concrete type is unnecessary, you can use the type wildcard `@_`.)

By default, the order of type variables that you assign types to is the same as the order in which they are introduced in the function definition.

Without `forall`, you would write something like this:

```Haskell
veryLongFunction :: a -> b -> c -> d -> e -> f -> g -> h -> i -> j
veryLongFunction = ...

func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
```
Quite long, right?
However, this can be simplified with explicit `forall`:

```Haskell
veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
veryLongFunction = ...

func = veryLongFunction @Integer ...
```

Since `j` is the first type variable in the declaration of `veryLongFunction`, you can specify only it.

> This is useful not only in such big examples but also when your function requires any of its arguments to be specified on usage.

Reviewer: Is is obvious what "on usage" means to the reader?

Reviewer: @PruStephan This is still unanswered.

Author: No it is not (At least I understood it only with third attempt). I thought I described it somewhere. Basically we are talking about on-usage (on-call) meaning for universal quantification and on-definition for existential types. On-usage means that when you call a function, you can specify type for it's arguments, like we did with TypeApplication. on-definition means that you can choose type for a existentially quantified variable when you define a function, and only there. Also, this concept looses meaning for existentially quantified datatypes

Reviewer 2: Agreed, I also didn't understand what that meant.
I ended up deleting that sentence in https://github.com/serokell/blog-posts/pull/4/commits/27ef8ba24b79f17e56a214f51fe7606d434821ae because I believe the same idea was already explained a few paragraphs above, where I explained `TypeApplications` and instantiation in more detail.

```haskell
{-# LANGUAGE ExplicitForAll #-}

id :: forall a. a -> a    
```

### Examples of usage

The syntax is simple – at the beginning of function or instance declaration, before any constraints or arguments are used, use
the `forall` quantifier to introduce all type variables that you will use later.

Here, we introduce four variables: `a`, `b`, `c`, and `m`:

```Haskell
func :: forall a b c m. a -> b -> m c -- OK, it compiles
```

All type variables that you introduce in your type signature should belong to exactly one `forall`. No more, no less.

```haskell
func :: forall a b c.   a -> b -> m c -- Error, type variable `m` is not introduced
```

Of course, we can also add constraints.
For example, we might want to specify that the `m` variable needs an instance of the `Monad` typeclass.

```Haskell
func :: forall a b c m. Monad m => a -> b -> m c
```
So far, it might seem not very useful.
Nothing changed when we added the quantifier because it was already there (although implicit).
However, making the quantification explicit allows us to do some new things.

### Practical use cases of universal quantification

> #### Reordering type variables

Reviewer: Doesn't this also fall under supporting other extensions?

Reviewer: Perhaps we can simplify it by just having a thesis of " explicit universal quantification is necessary for / helps to work with other extensions" and then list the examples of where it is used in extensions. 

It doesn't change the content, but the framing makes it much easier to understand what one would do with it.

### Supporting other extensions

On its own, `ExplicitForAll` doesn't do a lot.

But you frequently need it to enable other extensions. Some of them just won't work without the `forall` quantifier. :)

The most common one is `ScopedTypeVariables`, which allows you to pass a type variable from an outer function to one defined in a where clause.

The code below will not compile because the compiler thinks that the `a` in `example`'s signature and the `a` in `pair`'s are two different types.

```haskell
example :: a -> [a] -> [a]
example x rest = pair ++ rest
  where
    pair :: [a]
    pair = [x, x]
```

To fix it, you need to use the `ScopedTypeVariables` extension and also add a `forall` quantifier to the `example` function, since the extension won't work without it.

```haskell
example :: forall a. a -> [a] -> [a]
```
You might also notice how `forall` was useful with the `TypeApplication` extension.
Other extensions that somehow benefit from the usage of `ExplicitForAll` are `LiberalTypeSynonyms`, `RankNTypes`, and many more.

## Existential quantification

As I said earlier, besides universally quantified variables, Haskell also supports existential types.

> They are also created with the `forall` keyword.

Reviewer: Would put an example of an existential type here.

They can be implemented this way because these two constructions – `(exists x. p x) -> q` and `forall x. (p x -> q)` – are equivalent in terms of first-order predicate logic.
For a theoretical proof of this statement, you can check this [thread](https://stackoverflow.com/questions/10753073/whats-the-theoretical-basis-for-existential-types?rq=1).

> ### Usage

Reviewer: I think there should be two sections: existential quantification in datatypes, and existential quantification in functions instead of usage and heterogenous list.

Author: Yeah, that sounds better

In order to proceed further, we'll have to enable the `RankNTypes` extension. This lets us write higher-rank types.

<hr>

**Higher-rank types**

By default, all types are rank-1. In a rank-1 type, a `forall` cannot appear on the left of an arrow (keep in mind that the `->` operator has precedence over `.`)

For example, this function has a rank-1 type:

```haskell
f :: forall a b. a -> b -> a

-- Or, to make the precedence explicit:
f :: forall a b. (a -> b -> a)
```

With the `RankNTypes` extension, we can write rank-2 types – types where `forall` appears to the left of one arrow:

```haskell
f :: (forall a. Show a => a -> IO ()) -> IO ()

> ```

Reviewer: Read on this a little bit, and I think it is much simpler to just say that rank 2 types take as a parameter a rank 1 type (`forall a. Show a => a -> IO ()`), just like higher-order functions take as arguments regular functions. 

Would also show a rank 3 function to hammer it down.

Reviewer: @PruStephan This is still unanswered.

Reviewer 2: Yes, that's a good point!
These two definitions are indeed equivalent. [Source](https://en.wikibooks.org/wiki/Haskell/Polymorphism): 

> In general, a rank-n type is a function that has at least one rank-(n-1) argument but no arguments of any higher rank.

And this perspective may be more intuitive and easier to grasp, so I agree it may be better to put it that way.

Reviewer 2: Some more good examples from this [StackOverflow answer](https://stackoverflow.com/a/22362474/857807):

```
Rank 0: Int
Rank 1: forall a. a -> Int
Rank 2: (forall a. a -> Int) -> Int
Rank 3: ((forall a. a -> Int) -> Int) -> Int
```

Which seem to match the examples in SPJ et al's paper ["Practical type inference for arbitrary-rank types"](https://www.microsoft.com/en-us/research/publication/practical-type-inference-for-arbitrary-rank-types/?from=http%3A%2F%2Fresearch.microsoft.com%2Fusers%2Fsimonpj%2Fpapers%2Fhigher-rank%2F):

![image](https://user-images.githubusercontent.com/718166/172882726-3fca6a09-51f7-420f-8935-7f28ba978d30.png)

So, instead of saying "By default, all types are rank-1", we should really be saying that Haskell98 supports monomorphic types, which are rank-0, and rank-1 polymorphic types.

Reviewer 2: Done here: https://github.com/serokell/blog-posts/pull/4/commits/d50ff4f65468c46dc454159dfef9d95813832ea2

<hr>

Because `forall a` appears to the left of one arrow, `a` is said to be existentially quantified.
It is also true for every `forall` that is introduced to the left of an odd number of functional arrows.
But what does this mean?

> In contrast to the universal quantifier, the existential one means that the function is itself responsible for instantiating what a type variable should mean. So the result type of existentially quantified variable will become known when we address it at the 'definition-site' (i.e., in the function body).

Reviewer: Would have a decently-sized section on what is instantiation.

Reviewer 2: Done here: https://github.com/serokell/blog-posts/pull/4/commits/63b2e18f537a32f858d243ace6ccb563710a4095

This also means that no one can specify its type outside of function.

Let's take a closer look on practical examples.

We will start with functions:

> ```haskell

Reviewer: I think you can separate two parts: 

existential quantification in datatypes 
existential quantification in functions

Because we want `a` to be instantiated at `func`'s definition-site (rather than its use-site), we need `a` to be existentially quantified instead.

```haskell
{-# LANGUAGE RankNTypes #-}

func :: (forall a. Show a => a -> IO ()) -> IO ()
```
Now the compiler knows that no matter which function we pass to `func`, type variable `a` should be instantiated here, not outside.

With this change, we can instantiate `a` when we call log.

```haskell
func :: (forall a. Show a => a -> IO ()) -> IO ()
func log = do
  username <- getUsername
  log @String username

  userCount <- getUserCount
  log @Int userCount
```

Note, that this type applications is unnecessary.
Compiler is smart enough to choose appropriate types.

#### Heterogeneous lists

Now we understand that existential types allow you to introduce type variables without the necessity to specify them.
Let's take a look how this is done in datatypes.

You can introduce existentially quantified type variable in a datatype by using `forall` on the left side of the constructor name. 

> Existential types in datatypes allow you to have field in a datatype without specifying its type.

Reviewer: Would hammer down the difference between existential and regular data type. 

Such as: 

data Elem a = Elem a 

A data type that, given a type, will return a concrete data type wrapping that type. 

data Elem = forall a. Elem a 

A data type that ....

Reviewer: Perhaps showing differences in types can show you how it hides the type. 

e.g.

*Main> :t Elem
Elem :: a -> Elem a
*Main> :t Elem2
Elem2 :: a -> Elem2

Reviewer: <img width="541" alt="image" src="https://user-images.githubusercontent.com/35335196/167640036-5c1395d8-0f95-4e85-aa31-1ad4f1bebe88.png">

> ```haskell

Reviewer: <img width="611" alt="image" src="https://user-images.githubusercontent.com/35335196/167632545-2d622335-caa6-4822-92cb-ba2c0f2f9714.png">

I don't think you asked me to have either ExistentialQuantification or GADTs here, should probably mention that.

data Elem = forall a. Elem a

multitypedList :: [Elem]
multitypedList = [Elem "a", Elem 1, Elem (Just 5)]
```

This is a heterogeneous list that contains values of different types.
You can't do that with an universally quantified list because every element in that list would need to have the same type.

But, there are not many things you can do with elements of this list.
Since the type of elements is unknown at the use-site, you can't address it, nor can you pass it to functions that expect values of concrete type.
`Data.Typeable` module's functionality can help to circumvent this restriction. It allows you to cast one type (even unknown) to another one.

Adding constraints to the `a` variable may also improve the situation:

```haskell

> -- You need to enable either `ExistentialQuantification` or `GADTs` extension to use this syntax

Reviewer: Ah, here it is. Yeah, you need to warn about this faster. 😂

showFunc (x:xs) => (show x) ++ (showFunc xs)
```

A useful example of hiding type variables is the `SomeException` wrapper.

```haskell
data SomeException = forall a. Exception a => SomeException a
```

It allows you to catch exceptions of any type.
As long as you haven't specified its type (and it is not obvious to the compiler, for example, from the handler), all exceptions will be caught, wrapped in `SomeException`.

However, again, there is a limitation.
We know nothing about the inner exception.
We can, for example, print it, but sometimes we need more information.
And we can't use type inheritance to describe what type we want to get.

```haskell
-- Something like that won't work.
-- We know too little about type inside,
-- to be sure that it is 'MyException'.
myFromException :: SomeException -> MyException
myFromException (SomeException (ex :: MyException)) = ex
```

Luckily for users, module `Control.Exception` provides special function `fromException`, the purpose of which is to extract the inner value and cast to a specified type. But how that is done is a topic for a different article.

> ### Existential quantification in functions

Reviewer: I guess the correct phrase would be "function signatures"? @PruStephan 🤔

Author: Yep, seems so.

```

All exceptions, upon being thrown, are wrapped in `SomeException`.
If you want to catch all thrown exceptions, you can use `catchAll` to catch a `SomeException`.

When you pattern match on the `SomeException` constructor, its `Exception` instance is brought into scope.
`Exception` [implies `Typeable` and `Show`](https://hackage.haskell.org/package/base/docs/Control-Exception.html#t:Exception),
so those instances are also brought into scope.

```hs
exceptionHandler :: SomeException -> IO ()
exceptionHandler (SomeException (ex :: e)) =
  -- The following instances are in scope here:
  -- `Exception e`, `Typeable e`, `Show e`
  ...
```

These 3 instances are all the information we  have about `ex` at compile-time.
Notably, when pattern matching on `SomeException`, the concrete type of `ex` is not known at compile-time.

```hs
data MyException = MkMyException deriving Show
instance Exception MyException

exceptionHandler :: SomeException -> IO ()
exceptionHandler (SomeException (ex :: MyException)) = ...
  --                             ^^^^^^^^^^^^^^^^^
  -- • Couldn't match expected type ‘e’ with actual type ‘MyException’
```

> (Luckily, we do have a `Typeable` instance in scope, which lets us perform a runtime cast,

Author: Don't you think that mentioning runtime cast is something we should avoid, or at least provide a link too. Because in my opinion, runtime casts is an advanced technique, that the Reader might be not familiar with?

Reviewer: A link should be good. 👍

Reviewer 2: > Don't you think that mentioning runtime cast is something we should avoid, or at least provide a link too

I wanted to avoid it, but that might leave the reader thinking `SomeException` is not of much use :/ So leaving a small note at the end of the section seems reasonable, IMO.

Adding a link would definitely help! Would a link to `Data.Typeable.cast` suffice? https://hackage.haskell.org/package/base/docs/Data-Typeable.html#v:cast

> Because in my opinion, runtime casts is an advanced technique, that the Reader might be not familiar with?

Hmm I guess it depends on the point of view. Runtime casts are pretty common and routine in most typed languages. Especially in Java in C#, it's hard to do anything useful without resorting to runtime casts.

Yes, in Haskell, runtime casting is a more advanced technique because it's not embedded in the language's syntax. The type information is reified as `TypeRep` and carried via the `Typeable` typeclass, so you're right, I wouldn't expect the reader to be familiar with `Typeable`, but I think it's fair to expect them to know the concept of runtime casting in general.

Author: > Yes, in Haskell, runtime casting is a more advanced technique because it's not embedded in the language's syntax

That's what I meant

> but I think it's fair to expect them to know the concept of runtime casting in general.

That seems to be true. I almost forgot that in C++ and Java casts are common thing)

Reviewer 2: Added the link ^^
